src/aggregator/aggregator/generic_elem.go

```diff
@@ -282,55 +288,96 @@ func (e *GenericElem) Consume(
+		e.toConsume = append(e.toConsume, copiedValue)
+
+		if !expired {
+			// Keep item. Expired values are GC'd below after consuming.
+			e.values = append(e.values, value)
 		}
-		e.values = e.values[:n]
 	}
 	canCollect := len(e.values) == 0 && e.tombstoned
 	e.Unlock()
 
+	var (
+		cascadeDirty  bool
+		prevTimeNanos int64
+	)
 	// Process the aggregations that are ready for consumption.
 	for i := range e.toConsume {
+		expired := e.toConsume[i].onConsumeExpired
 		timeNanos := timestampNanosFn(e.toConsume[i].startAtNanos, resolution)
+		// seed the previous timestamp if this is first consumed value.
+		if prevTimeNanos == 0 {
+			for k := range e.consumedValues {
+				if k > prevTimeNanos && k < timeNanos {
+					prevTimeNanos = k
+				}
+			}
+		}
+
 		e.toConsume[i].lockedAgg.Lock()
-		e.processValueWithAggregationLock(
-			timeNanos,
-			e.toConsume[i].lockedAgg,
-			flushLocalFn,
-			flushForwardedFn,
-			resolution,
-		)
+
+		// if a previous timestamps was dirty, that value might impact a future derivative calculation, so
+		// cascade the dirty bit to all succeeding values. there is a check later to not resend a value if it doesn't
+		// change, so it's ok to optimistically mark dirty.
+		if cascadeDirty || e.toConsume[i].lockedAgg.dirty {
+			cascadeDirty = e.processValueWithAggregationLock(
+				timeNanos,
+				prevTimeNanos,
+				e.toConsume[i].lockedAgg,
+				flushLocalFn,
+				flushForwardedFn,
+				resolution,
+			)
+			e.toConsume[i].lockedAgg.flushed = true
+			e.toConsume[i].lockedAgg.dirty = false
+		}
+
 		// Closes the aggregation object after it's processed.
-		e.toConsume[i].lockedAgg.closed = true
-		e.toConsume[i].lockedAgg.aggregation.Close()
-		if e.toConsume[i].lockedAgg.sourcesSeen != nil {
-			e.cachedSourceSetsLock.Lock()
-			// This is to make sure there aren't too many cached source sets taking up
-			// too much space.
-			if len(e.cachedSourceSets) < e.opts.MaxNumCachedSourceSets() {
-				e.cachedSourceSets = append(e.cachedSourceSets, e.toConsume[i].lockedAgg.sourcesSeen)
+		if expired {
+			// Cleanup expired item.
+			e.toConsume[i].lockedAgg.closed = true
+			e.toConsume[i].lockedAgg.aggregation.Close()
+			if e.toConsume[i].lockedAgg.sourcesSeen != nil {
+				e.cachedSourceSetsLock.Lock()
+				// This is to make sure there aren't too many cached source sets taking up
+				// too much space.
+				if len(e.cachedSourceSets) < e.opts.MaxNumCachedSourceSets() {
+					e.cachedSourceSets = append(e.cachedSourceSets, e.toConsume[i].lockedAgg.sourcesSeen)
+				}
+				e.cachedSourceSetsLock.Unlock()
+				e.toConsume[i].lockedAgg.sourcesSeen = nil
 			}
-			e.cachedSourceSetsLock.Unlock()
-			e.toConsume[i].lockedAgg.sourcesSeen = nil
 		}
+
 		e.toConsume[i].lockedAgg.Unlock()
-		e.toConsume[i].Reset()
+		if expired {
+			e.toConsume[i].Release()
+			// the consumed value of the previous timestamp is no longer needed once this value has expired.
+			delete(e.consumedValues, prevTimeNanos)
```

Comment:
Hm, so this happens outside of `e.Lock()` and `e.Unlock()`. How do you feel about that? Wonder if we should reacquire lock or find some other way to reason about concurrent mutations to it. (it is read inside `processValueWithAggregationLock` yeah? I guess did we not lock on this before? ugh)

Reply:
it's safe to happen outside of the lock because only the flusher uses `consumedValues`, which is single threaded.

Reply:
the lock is only for state that incoming writes might mutate (like the aggregated value itself)

Reply:
Yeah, that's fair and agreed. The only other modification is on Close or something yeah?  I suppose we wouldn't try to be flushing and then Close it at the same time.

Reply:
so both `toConsume` and `consumedValues` are internal state only used by the flusher. the flusher first calls `Consume` and then calls `Close` for any elements that can be collected. no need to sync, since it's single threaded. I added some comments.